py/dml/messages.py

```diff
@@ -1138,6 +1138,25 @@ class EAUTOPARAM(DMLError):
     library, and they may not be overridden."""
     fmt = "bad declaration of automatic parameter '%s'"
 
+class ENWALRUS(DMLError):
+    """When the `walrus_params` provisional feature is enabled, parameter
+    declarations that do not override existing declarations must be
+    declared using the `:=` or `:default` syntax.
+    """
+    fmt = "non-overriding param should use :%s syntax (from walrus_params feature)"
+
+class EWALRUS(DMLError):
+    """When the `walrus_params` provisional feature is enabled, parameters
+    declared using `:=` or `:default` syntax may not override other parameters.
+    """
+    fmt = "use of :%s to override existing param (from walrus_params feature)"
```

Comment:
`(the :%s syntax is provided by the walrus_params provisional feature)`, possibly with `... provisional feature, which has been enabled)`

Reply:
By the way, I just realized that the following isn't illegal:
```
template x {
    param p = 4; // := with the new feature
}

template y is x {
    param p;
}

is y;
```
Perhaps it should be? I'm not sure. There's also the interplay with late shared, we don't want to make:
```
template x {
    param p = 4; // := with the new feature
}

template y is x {
    param p : int;
}

is y;
```
illegal.

Reply:
Errgh... late shared really does make this _weird_, doesn't it. Ideally we want `:=` to not be allowed to be used as an _override_, and yet...

```
template x {
    param p :default 4;
}

template y is x {
    param p : int;
    param p = 7;
}
```
Should be legal. If walruses truly are just syntactic sugar, why shouldn't
```
template y is x {
    param p : int = 7;
}
```
also be legal? And, by extension,

```
template y is x {
    param p := 7;
}
```
?

I think the answer should be that `:=` is only _mostly_ syntactic sugar, but does add an additional restriction that the multi-statement form does not: that the the param must not have been previously declared.

Reply:
Regarding your original comment: I chose instead to not mention the provisional feature. If the walrus provisional is enabled in a file and you write `:=` in that file, then you are aware of walruses, so that information is redundant. It would make more sense to add a walrus hint to the `ESYNTAX` for `:=`, we can do that when the provisional goes stable.